### PR TITLE
Outline comments

### PR DESCRIPTION
**Pull request by vasily-kirichenko**

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
"Comments enabled" setting works OK, but "Comments enabled" does not: open the dialog, uncheck it, close the dialog, open again - it's checked :( No idea why. Checked everything 5 times.

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 1]
```diff
@@ -1149,11 +1150,58 @@ module Outlining =
...
+    type private LineNum = int
+    type private LineStr = string
+    type private MultilineComment = (LineNum * LineStr) list
+
+    let getCommentRanges (lines: string[]) =
+        let comments: MultilineComment list =
+            lines
+            |> Array.foldi (fun ((lastLineNum, currentComment, comments) as state) lineNum lineStr ->
+                match lineStr.TrimStart() with
+                | x when x.StartsWith "//" ->
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: We have `StartsWith` active patterns in Core project.

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 2]
```diff
@@ -1149,11 +1150,58 @@ module Outlining =
               yield! collectOpens decls
               yield! Seq.collect parseDeclaration decls }
 
-    let getOutliningRanges tree =
+    type private LineNum = int
+    type private LineStr = string
+    type private MultilineComment = (LineNum * LineStr) list
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: Is this really `MultilineComment` (comments between (* and *) )? I think it's more like `CommentList`.

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 3]
```diff
@@ -1149,11 +1150,58 @@ module Outlining =
...
+                | _ -> lastComment :: comments
+                |> List.rev
+
+        comments
+        |> List.choose (fun comment ->
+            match comment with
+            | [] | [_] -> None
+            | _ ->
+                let comment = comment |> List.rev |> Array.ofList
+                let startLine, endLine = fst comment.[0], fst comment.[comment.Length - 1]
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: Maybe reorganize these a bit:

```fsharp
let startLine, startStr = comment.[0]
let endLine, endStr = comment.[comment.Length - 1]
let startCol = startStr.IndexOf '/'
let endCol = endStr.TrimEnd().Length
```

**Comment by dungpa** [comment]
Could you add outlining tests that contain comments? 

This PR doesn't recognize comments in a precise way e.g. having false positives on `//` inside multi-line strings and `//` inside block comments, but I think it's a very good start.

**Comment by vasily-kirichenko** [comment]
All done + a distinct category for XML doc comments.

But the bug with storing "Enabled" setting for the two new categories is still here - it cannot be set to "false": uncheck the checkbox, close the dialog, open it again, the checkbox is checked. This is reproduced for only two checkboxes - [x] Enable comments and [x] Enable XML doc comments. I've checked all the UI code again, twice, no results :(

**Comment by dungpa** [comment]
Should `_control.Comments.Enabled` be `_control.Comments.Enabled` in https://github.com/vasily-kirichenko/FSharpVSPowerTools/blob/b507b7ef338670cc141ba8dc06f3f4cebbfc1f4f/src/FSharpVSPowerTools/UI/OutliningOptionsPage.cs#L99-L105? The `Enabled` property is inherited from somewhere and it has a different purpose.

**Comment by vasily-kirichenko** [comment]
OMG, yes! Tested all combinations of comments settings - all works. Thanks a lot.

**Comment by dungpa** [comment]
:+1:

**Pull request merged**